### ISSUE TITLE
 Bug report: gpgtools hangs on initial install

### ISSUE THREAD

**Issue by Lavoaster** [caskroom/homebrew-cask#29297]
On a fresh installation of a gpg tools, where it has never touched the system before, `brew cask install gpgtools` installs gpg tools but never ends. 

I'm running macOS Sierra but it has affected El Capitan previously.

```
adam@legolas|~/macos
) brew cask install gpgtools
==> Caveats
Cask gpgtools installs files under "/usr/local". The presence of such
files can cause warnings when running "brew doctor", which is considered
to be a bug in Homebrew-Cask.

==> Downloading https://releases.gpgtools.org/GPG_Suite-2016.10_v2.dmg
######################################################################## 100.0%
==> Verifying checksum for Cask gpgtools
==> Running installer for gpgtools; your password may be necessary.
==> Package installers may write to any location; options such as --appdir are ignored.
Password:
==> installer: Package name is GPG Suite
==> installer: Installing at base path /
==> installer: The install was successful.
^C
-> [130]
adam@legolas|~/macos
) 
```

So here's the thing - You can't reproduce this on the same machine after it occurs. 

I haven't looked very far into how much cleanup is done, but I've ran the following tasks in my attempts to reproduce this on the same machine - `brew cask remove gpgtools` and `brew cask cleanup`.

My actions to try reproduce this have been
- `sudo -v` and then `brew cask install gpg-tools`
- No `sudo -v` (fresh terminal instance), only `brew cask install gpg-tools` (prompts for password)

Oh, I've also reproduced this in the following shell environments `bash` `zsh (omz)` `zsh (prezto)` if it matters to know about that.

If you need any more information from me, I'll be happy to provide it, but I am out of machines to test this on. If you have any advice on being able to bring up a disposable macOS machine, I'll be happy to debug it from there.

**Comment by vitorgalvao** [comment]
Are you sure it is hanging and not just taking a long time (rare, but happens with some casks)?

Next step would be to try installing the `.pkg` manually and trying to install it via the CLI (but without HBC). All on clean machines.

> If you have any advice on being able to bring up a disposable macOS machine

Best way would be to use a [VM](https://en.wikipedia.org/wiki/Virtual_machine). Problem is at the moment I don’t have access to one ([long story](https://github.com/caskroom/homebrew-cask/issues/26872#issuecomment-271308215), I’m dependent on [Veertu](https://veertu.com/)).

[Virtualbox](https://www.virtualbox.org/) is free and might work for you (last time I checked it didn’t work on my laptop).

**Comment by Lavoaster** [comment]
@vitorgalvao Yeah, I've previously had this run for quite some time (> 1 hour) unattended during our provision scripts at work until someone noticed and it was manually restarted.

I'll take a gander at getting a VM up and get back to here later today on if I was able to or not.

**Comment by miccal** [comment]
Bah, after all the effort to get the [`uninstall` to work](https://github.com/caskroom/homebrew-cask/issues/28911), now it won't install!

**Comment by Lavoaster** [comment]
@miccal To be clear, it does install. Something just doesn't end.

Anyway, to give an update I said I would give earlier. Apparently setting up an macOS VM isn't exactly the easiest thin in the world, I'll give it a few more attempts and get back when I've managed to get something.

**Comment by miccal** [comment]
No problem @Lavoaster, I was not frustrated at you!

Unfortunately I cannot help you test this as I have `gpg` installed via `brew` on all my machines.

**Comment by larrydalaw** [comment]
I have same problem too.. now i am waiting for a longer time to see anything happens...
after some waiting..decided not to , then i uninstall it... after 
Removing launchctl service on below : org.gpgtools.gpgmail.patch-uuid-user
service org.gpgtools.macgpg2.fix, service org.gpgtools.macgpg2.shutdown-gpg-agent ,service org.gpgtools.macgpg2.updater, org.gpgtools.macgpg2.gpg-agent
and 
==> Quitting application ID  on below : com.apple.mail
org.gpgtools.gpgkeychainaccess, org.gpgtools.gpgkeychain, org.gpgtools.gpgservices
it says 
==> Running uninstall script /usr/local/Caskroom/gpgtools/2016.10_v2/Uninstall.app/C
==> No matching processes were found
==> Forgot package 'org.gpgtools.checkprivatekey.pkg' on '/'. (and some other .pkg and .key .preinstall . finally removing some files and end..

the whole log is here 
`
s-MacBook-Pro:Sites $ brew cask install gpgtools
==> Caveats
Cask gpgtools installs files under "/usr/local". The presence of such
files can cause warnings when running "brew doctor", which is considered
to be a bug in Homebrew-Cask.

==> Downloading https://releases.gpgtools.org/GPG_Suite-2016.10_v2.dmg
######################################################################## 100.0%
==> Verifying checksum for Cask gpgtools
==> Running installer for gpgtools; your password may be necessary.
==> Package installers may write to any location; options such as --appdir are ignor
Password:
==> installer: Package name is GPG Suite
==> installer: Installing at base path /
==> installer: The install was successful.
Ω^Z
[1]+  Stopped                 brew cask install gpgtools
s-MacBook-Pro:Sites \$ brew cask uninstall gpgtools
==> Running uninstall process for gpgtools; your password may be necessary
==> Removing launchctl service org.gpgtools.Libmacgpg.xpc
Password:
==> Removing launchctl service org.gpgtools.gpgmail.patch-uuid-user
==> Removing launchctl service org.gpgtools.macgpg2.fix
==> Removing launchctl service org.gpgtools.macgpg2.shutdown-gpg-agent
==> Removing launchctl service org.gpgtools.macgpg2.updater
==> Removing launchctl service org.gpgtools.macgpg2.gpg-agent
==> Quitting application ID com.apple.mail
==> Quitting application ID org.gpgtools.gpgkeychainaccess
==> Quitting application ID org.gpgtools.gpgkeychain
==> Quitting application ID org.gpgtools.gpgservices
==> Running uninstall script /usr/local/Caskroom/gpgtools/2016.10_v2/Uninstall.app/C
==> No matching processes were found
==> Forgot package 'org.gpgtools.checkprivatekey.pkg' on '/'.
==> 
==> Forgot package 'org.gpgtools.gpgkeychain.pkg' on '/'.
==> 
==> Forgot package 'org.gpgtools.gpgpreferences.pkg' on '/'.
==> 
==> Forgot package 'org.gpgtools.gpgservices.pkg' on '/'.
==> 
==> Forgot package 'org.gpgtools.key' on '/'.
==> 
==> Forgot package 'org.gpgtools.libmacgpg.xpc.pkg' on '/'.
==> 
==> Forgot package 'org.gpgtools.libmacgpgB.pkg' on '/'.
==> 
==> Forgot package 'org.gpgtools.macgpg2.pkg' on '/'.
==> 
==> Forgot package 'org.gpgtools.pinentry.pkg' on '/'.
==> 
==> Forgot package 'org.gpgtools.pkg.gpgmail' on '/'.
==> 
==> Forgot package 'org.gpgtools.pkg.preinstall' on '/'.
==> 
==> Forgot package 'org.gpgtools.updater.pkg' on '/'.
==> 
==> Removing files from pkgutil Bill-of-Materials
==> Removing files: ["/Library/Services/GPGServices.service", "/Library/Mail/Bundles
s-MacBook-Pro:Sites $ `

**Comment by commitay** [comment]
I believe I've found the cause for this.

Testing done in a new vm with 10.12.4 and CLI tools 8.3.


Manual `.pkg` (gui) and `sudo installer -pkg` (cli) both work fine. (Without HB installed.)


`brew cask install gpgtools` (fresh HB with nothing installed.)

Let this run for a couple of hours to see if it finished. Stuck at the end of install.


Tried removing the `postflight` command from the cask and it installs properly.

```
  # TODO: remove all ENV variables
  postflight do
    system_command '/usr/local/MacGPG2/libexec/fixGpgHome',
                   args: [
                           Etc.getpwuid(Process.euid).name,
                           ENV['GNUPGHOME'] ? ENV['GNUPGHOME'] : Pathname.new(File.expand_path('~')).join('.gnupg'),
                         ],
                   sudo: true
  end
```

The postflight was added in #3366 to fix #3360.

> After review, I consider this to be an upstream bug, but we will do our best to provide a workaround in homebrew-cask.

The ownership issue in #3360 *seems* have to been resolved upstream. The owner of `~/.gnupg` and files in the folder after install (for me) is the user.

I'm not sure if I sure should do a PR for this without someone else confirming this first.

**Comment by miccal** [comment]
Thank you @commitay.
